### PR TITLE
more precise type facts for intersection

### LINKED ISSUE #45801: Type narrowing based on function typeof is broken when function is intersected with a record
# Bug Report

### 🔎 Search Terms

type narrowing based on function typeof is broken when intersection with a record is used

### 🕗 Version & Regression Information

- Recent change in behaviour happened in version 4.3.5

### ⏯ Playground Link
 
[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.2#code/C4TwDgpgBAshwEMoF4oG8oDMD22BcUAzsAE4CWAdgOZQC+AUJcBCZggMbQBi6UAFAEoCxctQb122CsSgAPFPwQE+PAGSx4CAVAA+RUpSrbkAPnT0olqGUz9QkbLaTIXUAESYArhXbAyUtwE0CytQgHowqGBCADoANxZCfwooE1QAFhiAZhiAVgIKCASSKABaUqhvCFlIXwgAExDQywio2OKkqSgAHgyYgCZsgjUNRDKK6trmRuarBABuJrooCAAbQmhg2ZbIyYg6+uEDaiXwyOj4xOTUvpz8-VEacpWa-enTq1aLjuveqEzBlkjo9dPwRnBENpnlVXgcPpYFksGLQgA)

### 💻 Code

```ts
type Meta = { foo: string }
interface F { (): string}
const x = (a: (F & Meta) | string) => {
    if (typeof a === "function"){
        // ts.version >= 4.3.5: never -- unexpected
        // ts.version <= 4.2.3: F & Meta -- expected
        a;
    } else {
        // expected: string
        // ts.version >= 4.3.5: string -- expected
        // ts.version <= 4.2.3: string | (F & Meta) -- unexpected
        a;
    }
}
```
### 🙁 Actual behavior

behavior was half correct from in TS versions  <= 4.2.3, but then it was changed in 4.3.5 which fixed the part that was wrong previously, but broke the other part that was correct. So we still have half broken behavior.

### 🙂 Expected behavior

It's expected to pick the correct parts from the both versions.

### PR DESCRIPTION
**Pull request by gabritto**
Fixes #45801

### Brief background
When narrowing a type, we rely on `TypeFacts` flags that tell us what narrowing facts could be true for a value of a given type. The flags cover the possible results of narrowing using `typeof` (e.g. `TypeFacts.TypeofEQString`), comparsions to `null` or `undefined` (e.g. `TypeFacts.UndefinedOrNull` means `if (v === null || v === undefined)` could be true), or if the value is truthy or falsy (e.g. `TypeFacts.Truthy` means `if (v) { ... }` could be true).
To compute the type facts for a type `T`, we call `getTypeFacts(T)`.

###  The issue
Up until v4.2, when computing type facts for an intersection type, we `or`ed the facts of each of the intersection type's component type. e.g. `getTypeFacts(A & B)` = `getTypeFacts(A) | getTypeFacts(B)`.
Starting at v4.3, [this changed to an `and`](https://github.com/microsoft/TypeScript/pull/43131/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R21763) of the facts of each intersection type's component type, e.g. `getTypeFacts(A & B)` = `getTypeFacts(A) & getTypeFacts(B)`, to fix cases like the following:

```ts
function f3<T>(x: T & string | T & number) {
    if (typeof x === "string") {
        x;  // Should narrow to T & string
            // <= 4.2: x: T & string | T & number
    }
    else {
        x;  // Should narrow to T & number
            // <= 4.2: x: T & string | T & number
    }
}
```

If we have `T & number`, we know the value can only be of type number (or null or undefined, in non-strict), never of type string. However, we don't know anything about `T`, so when computing type facts for `T`, we think it can be anything at all (including of type string). If we `or` the type facts for `T` and `number`, then, we get that `T & number` could be string. If we `and` the type facts, we know `T & number` can't be of type string, because `number` can't be of type string.

However, in some cases we don't want an `and` of the type facts. Consider the example from issue #45801:

```ts
type Meta = { foo: string }
interface F { (): string}
const x = (a: (F & Meta) | string) => {
    if (typeof a === "function"){
        // ts.version >= 4.3.5: never -- unexpected
        // ts.version <= 4.2.3: F & Meta -- expected
        a;
    } else {
        // ts.version >= 4.3.5: string -- expected
        // ts.version <= 4.2.3: string | (F & Meta) -- unexpected
        a;
    }
}
```

For type `F & Meta`, we know values of `F` have to be functions, since `F` has a call signature declared. But type `Meta` doesn't, it is a regular object. If we `and` those facts, we conclude `F & Meta` cannot be a function, because we don't think values of `Meta` could be functions (no call signature). So in this case, we probably want to `or` the type facts.

(Note: We could include `TypeFacts.TypeofEQFunction` in the flags returned for object types such as `Meta` above, and keep using `and` for computing type facts of intersection types. I think we don't want to do that, since we explicitly make that distinction in `getTypeFacts`'s code and it generally seems a useful distinction.)

### The possible fix

Considering both examples above, it seems we sometimes know a fact will *always* be true (i.e. true for all values of a type), e.g. values of type `T & number` can't be strings (`if (typeof v === 'string')` can't be true). This contrasts with the other uses of type facts, which are facts that *can* be true for values of type T, but are not necessarily always true/true for all values.
So that's why sometimes we want to use an `and` to compute facts of intersection types, and sometimes we want to use an `or`.

#### *tl;dr*
This PR changes the way we compute type facts of intersection types. For each component type, in addition to computing its type facts, we also compute which facts are always true (true for all values of that type) and which are always false.
We then `or` the regular type facts of the component types, and we also `or` the always false facts. We then exclude the facts that are always false from the regular facts.
So, in cases like `T & number`, even though when we `or` the components' type facts, we include a flag such as `TypeFacts.TypeofEQString` (because that's one of the flags for `T`), we will delete it later because `TypeFacts.TypeofEQString` is always false for type `number`.

#### Questions I had
- ~~When is `TypeFacts.TypeofEQHostObject` always false?~~ It seems `typeof v === "xxx"` ("xxx" being something other than the usual strings returned by `typeof`) can only be true if `v` is a host object (and hosts return a custom `typeof` for those), or if we don't yet recognize the "xxx" in question (e.g. if there's ever a new `Fraction` type and `typeof v === "fraction"`).  So it should always be false for known primitives at least.
- ~~Should we care about contradictory types (e.g. `string & number`) in `getTypeFacts` or `getAlwaysFalseTypeFacts`?~~ I'm assuming we shouldn't, because existing code for `getTypeFacts` doesn't address this, and this looks like something `getIntersectionType` takes care of.

### CODE REVIEW

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot run dt

**Comment by andrewbranch** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot perf test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -22935,7 +23023,10 @@ namespace ts {
                     (type === falseType || type === regularFalseType) ? TypeFacts.FalseStrictFacts : TypeFacts.TrueStrictFacts :
                     (type === falseType || type === regularFalseType) ? TypeFacts.FalseFacts : TypeFacts.TrueFacts;
             }
-            if (flags & TypeFlags.Object && !ignoreObjects) {
+            if (flags & TypeFlags.Object) {
+                if (ignoreObjects) {
+                    return TypeFacts.None;
```
Comment (gabritto) [inline src/compiler/checker.ts:23028]: `ignoreObjects` is only going to be true if we are in the middle of computing type facts for an intersection type. In this case, the neutral value returned by `getTypeFacts` can't be `TypeFacts.All` (as it was before), because we are `or`ing the type facts. So it should be `TypeFacts.None`.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -22964,13 +23055,177 @@ namespace ts {
...
+                alwaysTrue |= tTrue;
+                alwaysFalse |= tFalse;
+            }
+
+            if (alwaysTrue & alwaysFalse) {
+                // Contradiction
+                return TypeFacts.None;
+            }
+
+            return (facts | alwaysTrue) & (~alwaysFalse);
```
Comment (gabritto) [inline src/compiler/checker.ts:23083]: I think we don't need to compute `alwaysTrue` facts, since they are supposed to be a subset of `facts`. So right now they're just being used for detecting contradictions, as above, which I want to get rid of.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -22964,13 +23055,177 @@ namespace ts {
...
+            for (const t of type.types) {
+                facts |= getTypeFacts(t, ignoreObjects);
+                const { alwaysTrue: tTrue, alwaysFalse: tFalse } = getAlwaysTypeFacts(t, ignoreObjects);
+                alwaysTrue |= tTrue;
+                alwaysFalse |= tFalse;
+            }
+
+            if (alwaysTrue & alwaysFalse) {
+                // Contradiction
+                return TypeFacts.None;
```
Comment (gabritto) [inline src/compiler/checker.ts:23080]: I'm not entirely sure that can happen. If we have a contradiction in an intersection type (e.g. `number & string`), would this type be reduced to `never` before it gets to narrowing/calling `getTypeFacts`?

Comment (RyanCavanaugh) [reply src/compiler/checker.ts:23080]: I don't think it can happen, based on reading `getIntersectionType`

**Comment by ahejlsberg** [comment]
I'm definitely in favor of fixing this issue, but I'm not crazy about adding 175 lines of code to do it. Couldn't we just say that certain type facts for intersections are computed as `and` and others are computed as `or`, but otherwise use the existing single function? For example, `TypeFacts.TypeofEQString` would be an `and` whereas `TypeFacts.TypeofEQFunction` would be an `or`.
